File: lib/features/parent_home/domain/entities/parent.dart
// Parent domain entity
// Moved from lib/models/parent.dart for Clean Architecture

import 'package:flutter/material.dart';

/// V2: Parent permission level enum
enum ParentPermission {
  viewOnly,       // 열람만 가능
  managePayments, // 결제 관리
  manageLessons,  // 레슨 관리 (일정 변경 등)
  fullAccess;     // 전체 권한

  String get label {
    switch (this) {
      case ParentPermission.viewOnly:
        return '열람 전용';
      case ParentPermission.managePayments:
        return '결제 관리';
      case ParentPermission.manageLessons:
        return '레슨 관리';
      case ParentPermission.fullAccess:
        return '전체 권한';
    }
  }

  /// Check if permission allows payments
  bool get canManagePayments =>
      this == ParentPermission.managePayments ||
      this == ParentPermission.fullAccess;

  /// Check if permission allows lesson management
  bool get canManageLessons =>
      this == ParentPermission.manageLessons ||
      this == ParentPermission.fullAccess;
}

/// Parent registration status enum
enum ParentStatus {
  pending, // Invitation sent, not yet registered
  active, // Registered and active
  inactive; // Deactivated

  String get label {
    switch (this) {
      case ParentStatus.pending:
        return '초대 대기';
      case ParentStatus.active:
        return '활성';
      case ParentStatus.inactive:
        return '비활성';
    }
  }

  Color get color {
    switch (this) {
      case ParentStatus.pending:
        return const Color(0xFFFF9800); // Orange
      case ParentStatus.active:
        return const Color(0xFF4CAF50); // Green
      case ParentStatus.inactive:
        return const Color(0xFF9E9E9E); // Grey
    }
  }
}

/// Invitation source enum (who invited the parent)
enum InvitationSource {
  student, // Invited by student
  teacher; // Invited by teacher

  String get label {
    switch (this) {
      case InvitationSource.student:
        return '학생 초대';
      case InvitationSource.teacher:
        return '선생님 초대';
    }
  }
}

/// Parent model
class Parent {
  final String id;
  final String userId; // User table FK (for auth)
  final String name;
  final String phone;
  final String? email;
  final String? profileImageUrl;
  final Color profileColor;
  final ParentStatus status;
  final DateTime createdAt;
  final DateTime? updatedAt;

  const Parent({
    required this.id,
    required this.userId,
    required this.name,
    required this.phone,
    this.email,
    this.profileImageUrl,
    required this.profileColor,
    this.status = ParentStatus.pending,
    required this.createdAt,
    this.updatedAt,
  });

  /// Get first character of name for avatar
  String get initial => name.isNotEmpty ? name[0] : '?';

  /// Check if parent is active
  bool get isActive => status == ParentStatus.active;

  /// Format phone number for display
  String get formattedPhone {
    if (phone.length == 11) {
      return '${phone.substring(0, 3)}-${phone.substring(3, 7)}-${phone.substring(7)}';
    }
    return phone;
  }

  /// Copy with new values
  Parent copyWith({
    String? id,
    String? userId,
    String? name,
    String? phone,
    String? email,
    String? profileImageUrl,
    Color? profileColor,
    ParentStatus? status,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Parent(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      name: name ?? this.name,
      phone: phone ?? this.phone,
      email: email ?? this.email,
      profileImageUrl: profileImageUrl ?? this.profileImageUrl,
      profileColor: profileColor ?? this.profileColor,
      status: status ?? this.status,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Parent && runtimeType == other.runtimeType && id == other.id;

  @override
  int get hashCode => id.hashCode;
}

/// Parent invitation model (for tracking invitations)
class ParentInvitation {
  final String id;
  final String studentId;
  final String? teacherId; // null if student invited
  final InvitationSource source;
  final String parentPhone;
  final String? parentEmail;
  final String invitationCode;
  final DateTime expiresAt;
  final bool isUsed;
  final DateTime createdAt;

  const ParentInvitation({
    required this.id,
    required this.studentId,
    this.teacherId,
    required this.source,
    required this.parentPhone,
    this.parentEmail,
    required this.invitationCode,
    required this.expiresAt,
    this.isUsed = false,
    required this.createdAt,
  });

  /// Check if invitation is expired
  bool get isExpired => DateTime.now().isAfter(expiresAt);

  /// Check if invitation is valid (not used and not expired)
  bool get isValid => !isUsed && !isExpired;

  /// Get remaining time until expiration
  Duration get remainingTime => expiresAt.difference(DateTime.now());

  ParentInvitation copyWith({
    String? id,
    String? studentId,
    String? teacherId,
    InvitationSource? source,
    String? parentPhone,
    String? parentEmail,
    String? invitationCode,
    DateTime? expiresAt,
    bool? isUsed,
    DateTime? createdAt,
  }) {
    return ParentInvitation(
      id: id ?? this.id,
      studentId: studentId ?? this.studentId,
      teacherId: teacherId ?? this.teacherId,
      source: source ?? this.source,
      parentPhone: parentPhone ?? this.parentPhone,
      parentEmail: parentEmail ?? this.parentEmail,
      invitationCode: invitationCode ?? this.invitationCode,
      expiresAt: expiresAt ?? this.expiresAt,
      isUsed: isUsed ?? this.isUsed,
      createdAt: createdAt ?? this.createdAt,
    );
  }
}

/// V2: Parent-Teacher direct connection model
///
/// Represents a direct connection between parent and teacher.
/// Used when parent registers child via teacher invitation.
class ParentTeacherConnection {
  final String id;
  final String parentId;
  final String teacherId;
  final String? studentId; // Connected child (null if not yet specified)
  final ParentPermission permission;
  final DateTime connectedAt;
  final DateTime? updatedAt;

  const ParentTeacherConnection({
    required this.id,
    required this.parentId,
    required this.teacherId,
    this.studentId,
    this.permission = ParentPermission.viewOnly,
    required this.connectedAt,
    this.updatedAt,
  });

  /// Check if child is connected
  bool get hasChild => studentId != null;

  /// Copy with new values
  ParentTeacherConnection copyWith({
    String? id,
    String? parentId,
    String? teacherId,
    String? studentId,
    ParentPermission? permission,
    DateTime? connectedAt,
    DateTime? updatedAt,
  }) {
    return ParentTeacherConnection(
      id: id ?? this.id,
      parentId: parentId ?? this.parentId,
      teacherId: teacherId ?? this.teacherId,
      studentId: studentId ?? this.studentId,
      permission: permission ?? this.permission,
      connectedAt: connectedAt ?? this.connectedAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is ParentTeacherConnection &&
          runtimeType == other.runtimeType &&
          id == other.id;

  @override
  int get hashCode => id.hashCode;
}

File: backend/app/models/parent.py
from datetime import datetime
from enum import StrEnum

from sqlalchemy import Boolean, DateTime, Enum, ForeignKey, Index, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, TimestampMixin, UUIDMixin


class ParentStatus(StrEnum):
    pending = "pending"
    active = "active"
    inactive = "inactive"


class ParentPermission(StrEnum):
    viewOnly = "viewOnly"  # noqa: N815
    managePayments = "managePayments"  # noqa: N815
    manageLessons = "manageLessons"  # noqa: N815
    fullAccess = "fullAccess"  # noqa: N815


class ParentChildRelationStatus(StrEnum):
    pending = "pending"
    active = "active"
    inactive = "inactive"


class ParentInvitationSource(StrEnum):
    student = "student"
    teacher = "teacher"


class Parent(UUIDMixin, TimestampMixin, Base):
    """Parent profile linked to a user account."""

    __tablename__ = "parents"

    user_id: Mapped[str] = mapped_column(String(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    phone: Mapped[str | None] = mapped_column(String(20), nullable=True)
    email: Mapped[str | None] = mapped_column(String(255), nullable=True)
    profile_image_url: Mapped[str | None] = mapped_column(Text, nullable=True)
    profile_color: Mapped[str | None] = mapped_column(String(7), nullable=True, default="#6B5B95")
    status: Mapped[ParentStatus] = mapped_column(
        Enum(ParentStatus, native_enum=True),
        nullable=False,
        default=ParentStatus.pending,
    )

    __table_args__ = (
        Index("uk_parent_user_id", "user_id", unique=True),
    )


class ParentChildRelation(UUIDMixin, Base):
    """Relationship between parent and student (child)."""

    __tablename__ = "parent_child_relations"

    parent_id: Mapped[str] = mapped_column(String(36), nullable=False)
    student_id: Mapped[str] = mapped_column(String(36), nullable=False)
    is_primary_guardian: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    is_billing_target: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    status: Mapped[ParentChildRelationStatus] = mapped_column(
        Enum(ParentChildRelationStatus, native_enum=True),
        nullable=False,
        default=ParentChildRelationStatus.active,
    )
    permission: Mapped[ParentPermission] = mapped_column(
        Enum(ParentPermission, native_enum=True),
        nullable=False,
        default=ParentPermission.viewOnly,
    )
    connected_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )
    unlinked_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    __table_args__ = (
        Index("uk_parent_child", "parent_id", "student_id", unique=True),
    )


class ParentInvitation(UUIDMixin, Base):
    """Invitation for linking a parent to a student."""

    __tablename__ = "parent_invitations"

    student_id: Mapped[str] = mapped_column(String(36), nullable=False)
    teacher_id: Mapped[str | None] = mapped_column(String(36), nullable=True)
    source: Mapped[ParentInvitationSource] = mapped_column(
        Enum(ParentInvitationSource, native_enum=True),
        nullable=False,
    )
    parent_phone: Mapped[str] = mapped_column(String(20), nullable=False)
    parent_email: Mapped[str | None] = mapped_column(String(255), nullable=True)
    invitation_code: Mapped[str] = mapped_column(String(20), nullable=False)
    expires_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    is_used: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )

    __table_args__ = (
        Index("uk_parent_invitation_code", "invitation_code", unique=True),
        Index("idx_parent_invitation_student", "student_id"),
    )


class ParentTeacherConnection(UUIDMixin, Base):
    """Connection between parent and teacher (via a student)."""

    __tablename__ = "parent_teacher_connections"

    parent_id: Mapped[str] = mapped_column(String(36), nullable=False)
    teacher_id: Mapped[str] = mapped_column(String(36), nullable=False)
    student_id: Mapped[str | None] = mapped_column(String(36), nullable=True)
    permission: Mapped[ParentPermission] = mapped_column(
        Enum(ParentPermission, native_enum=True),
        nullable=False,
        default=ParentPermission.viewOnly,
    )
    connected_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

    __table_args__ = (
        Index("uk_parent_teacher", "parent_id", "teacher_id", unique=True),
        Index("idx_ptc_teacher", "teacher_id"),
    )


class ParentVisibilitySettings(UUIDMixin, TimestampMixin, Base):
    """Teacher-controlled parent visibility settings for a student."""

    __tablename__ = "parent_visibility_settings"

    teacher_id: Mapped[str] = mapped_column(String(36), nullable=False)
    student_id: Mapped[str] = mapped_column(String(36), nullable=False)
    can_view_schedule: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    can_view_assignments: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    can_view_practice: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    can_view_lesson_notes: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    can_view_recordings: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    can_view_detailed_feedback: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    can_view_chat: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)

    __table_args__ = (
        Index("uk_parent_visibility_teacher_student", "teacher_id", "student_id", unique=True),
        Index("idx_parent_visibility_student", "student_id"),
    )
